conbench/entities/benchmark_result.py

```diff
@@ -108,174 +158,259 @@ class BenchmarkResult(Base, EntityMixin):
+                raise BenchmarkResultValidationError(
+                    f'iterations count ({ stats_usergiven["iterations"] }) does '
+                    f"not match sample count ({len(samples)})"
+                )
+
+        if len(samples) >= 3:
+            # See https://github.com/conbench/conbench/issues/802 and
+            # https://github.com/conbench/conbench/issues/1118
+            q1, q3 = np.percentile(samples, [25, 75])
+            aggregates = {
+                "q1": q1,
+                "q3": q3,
+                "mean": np.mean(samples),
+                "median": np.median(samples),
+                "min": np.min(samples),
+                "max": np.max(samples),
+                # https://numpy.org/doc/stable/reference/generated/numpy.std.html
+                # This has N in the divisor (ddof=0)
+                "stdev": np.std(samples),
+                "iqr": q3 - q1,
+            }
 
-        # create if not exists
-        c = {"name": name, "tags": tags}
-        case = Case.first(**c)
+            # Now, overwrite with self-derived aggregates:
+            for key, value in aggregates.items():
+                result_data_for_db[key] = sigfig.round(value, sigfigs=5)
+
+                # Log upon conflict. Lett the automatically derived value win,
+                # to achieve consistency between the provided samples and the
+                # aggregates.
+                if key in stats_usergiven:
+                    if stats_usergiven[key] != value:
+                        log.warning(
+                            "key %s, user-given val %s vs. calculated %s",
+                            key,
+                            stats_usergiven[key],
+                            value,
+                        )
+
+        if len(samples) == 1:
+            if stats_usergiven["iterations"] == 1:
+                # If user provides aggregates, then it's unclear what they
+                # mean.
+                for k in ("q1", "q2", "mean", "median", "min", "max", "stdev", "iqr"):
+                    if stats_usergiven.get(k) is not None:
+                        raise BenchmarkResultValidationError(
+                            f"one data point from one iteration: property `{k}` "
+                            "is unexpected"
+                        )
+
+            if stats_usergiven["iterations"] > 1:
+                # Note(JP): I think that this here is precisely the one way to
+                # report the result of a microbenchmark with cross-repetition
+                # stats: data: length 1 (the duration of running many
+                # repetitions), and iterations: a number larger than 1, while
+                # mean, max, are now set (then these can be assumed to be
+                # derived from _within_ the microbenchmark. Note that this is a
+                # very special condition). The mean time for a single
+                # repetition now is sample/iterations.
```

Comment:
The issue with this is the frameworks that do this (googlebench, folly) can also run multiple repetitions, which would have to get reported like multisamples, so `iterations` etc would mean two different things for those cases. Maybe eventually we could accept per-iteration (read: repetition) stats?

Reply:
Thanks! New PR with this code section is here #1119, and I've linked your comment from here: https://github.com/conbench/conbench/pull/1119#discussion_r1167300853